libtcmu.c

```diff
@@ -969,17 +969,21 @@ struct tcmulib_cmd *tcmulib_get_next_command(struct tcmu_device *dev)
 
 			if (cdb_len < 0) {
 				/*
-				 * This should never happen so just drop cmd
+				 * This can only happen if a vendor-specific
+				 * command is received, so just drop cmd
 				 * for now instead of adding a lock in the
 				 * main IO path.
 				 */
-				break;
+				TCMU_UPDATE_DEV_TAIL(dev, mb, ent);
+				return NULL;
 			}
```

Comment:
I need a little extra info on how this helps.

It looks like without the patch we hit the break, do TCMU_UPDATE_DEV_TAIL at the bottom of the loop and either return NULL like above or try to handle another command. Either way, the vendor command will be dropped. The initaitor side will timeout, send TMFs which LIO basically just sits on, and then the initiator will offline the device.

With the patch it looks like we will do the update, and return NULL right away. The command will still be dropped and the same thing will happen on the initaitor side.

Since I was completely wrong about this never happening we should probably fix this properly. How about the attached patch which is not tested. It will handle the vendor command case. For the memory allocation failure there is nothing we can do and its going to hang. We could add some locking around the parts of tcmulib_command_complete that interact with the ring and then put that in a new function and call it, but I'm not sure that is worth it.
[runner-force-vendor-cdb-fail.txt](https://github.com/open-iscsi/tcmu-runner/files/1782406/runner-force-vendor-cdb-fail.txt)

Reply:
One more question.

How did you hit the vendor command case? Is it a normal case for some distro/OS or did you use something like sg utils to inject them?

Reply:
Maybe it's just better to move the tcmur device mbox lock to a common place. Attached is a patch which should handle all cases for all handler types. It is not tested.

[runner-make-mbox-lock-common.txt](https://github.com/open-iscsi/tcmu-runner/files/1782485/runner-make-mbox-lock-common.txt)

Reply:
Hi Mike,

As Damien mentioned, I was experimenting with a handler explicitly designed to handle some vendor-specific commands. I observed that with my patch TMF will not get stuck and the device is not offlined. That saved me a reboot, but, the best solution would be to just fail the command.

It should be possible to add the capability to handle vendor-specific commands to tcmu-runner by adding a new "get_cdb_length()" function to handler template allowing a particular backend handler to override the default tcmu_get_cdb_lba() function. I am not sure if such a feature is useful enough to implement.

Anyway, I tried your patches and runner-force-vendor-cdb-fail.txt works ok. The vendor command is failed right away. There is a slight problem - two different "command not supported" messages are logged, one for the actual opcode and the other for 0xc0, which might be a bit confusing for the user.

The second patch, runner-make-mbox-lock-common.txt, doesn't work right away in my environment. SCSI sd device does not appear in the system and the processing thread sleeps in ppoll() in the main loop. This happens after the first incoming command is completed. I can't see any obvious problem with the patch or any lockup happening. I am going to investigate further what is going on and will let you know. I think this patch is the best solution to this whole thing and also a nice cleanup.

Reply:
> Agree on the second patch being best. I will debug it real quick here too.

> It should be possible to add the capability to handle vendor-specific commands to tcmu-runner by adding a new "get_cdb_length()" function to handler template allowing a particular backend handler to override the default tcmu_get_cdb_lba() function. I am not sure if such a feature is useful enough to implement.

I am ok with adding vendor command support too. The override idea is fine with me.

Reply:
Ok.  I found the bug. It works for me because I somehow get multiple commands. If you were to only get one when this happens then we never called tcmulib_processing_complete and so the kernel is never notified of the completed failed command and it is never completed in the kernel.

Here is a updated patch.
[runner-make-mbox-lock-common-v2.txt](https://github.com/open-iscsi/tcmu-runner/files/1786438/runner-make-mbox-lock-common-v2.txt)

Reply:
Hey,

I made one other tiny change to the v2 patch and made a PR

https://github.com/open-iscsi/tcmu-runner/pull/377

We can discuss this specific issue there.